Add support for 4 MPTT inverter

mletenay opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Inverter "Operation Mode" select values were changed to snake_case and localized.
"General mode" -> "general"
"Off grid mode" -> "off_grid"
"Backup mode" -> "backup"
"Eco mode" -> "eco"
"Peak shaving" -> "peak_shaving"
"Eco charge mode" -> "eco_charge"
"Eco discharge mode" -> "eco_discharge"


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
- Add support for GE inverters with 4 MPTT strings.
- Upgrade of underlying goodwe library v0.2.25.
- https://github.com/marcelblijleven/goodwe/compare/v0.2.18...v0.2.25

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #86497
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

mletenay committed:
Add support for 4 MPTT inverter

starkillerOG reviewed (approved):
Looks good to me, thanks @mletenay

mletenay commented:
@starkillerOG Do you happen to know how to make the review/submit process faster ?
This one is kinda blocking fixes to #87477 #81601, since the underlying library has to be upgraded there again ...

kukulich commented:
Just tested and it looks good 👍

starkillerOG commented:
@mletenay please first resolve the conflicts with the main branch.
Then I can tag someone from the core-team that can merge this.

mletenay committed:
Merge branch 'dev' into goodwe86497

mletenay commented:
@starkillerOG Done, thx.

starkillerOG commented:
@MartinHjelmare would you be so kind to review this, I and @kukulich have already looked at it and approved.

MartinHjelmare commented:
Please don't ping for review unless I'm already involved. Thanks!

mletenay committed:
Merge branch 'dev' into goodwe86497

mletenay committed:
Fix import order after conflict merge

mletenay committed:
Update to goodwe 0.2.25 and its API changes

mletenay commented:
@starkillerOG FYI - I have bumped up the version to latest 0.2.25 (and adjusted code to its API changes) and moved sensor changes out of this PR to PR #87914

MartinHjelmare commented on homeassistant/components/goodwe/select.py:
```diff
@@ -38,24 +47,27 @@ async def async_setup_entry(
     device_info = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE_INFO]
 
+    supported_modes = await inverter.get_operation_modes(False)
+    OPERATION_MODE.options = [
```
Don't modify the constant description. Pass the discovered modes separately to the entity.

mletenay replied:
Done.

MartinHjelmare commented on homeassistant/components/goodwe/select.py:
```diff
@@ -15,17 +15,26 @@
+}
+
+
+def _get_operation_mode(mode: str) -> OperationMode:
```
Make another dict instead that's the inversion of the dict above.

mletenay replied:
Done, inverter dict created.

MartinHjelmare commented on homeassistant/components/goodwe/select.py:
```diff
@@ -15,17 +15,26 @@
 
 
+_OPERATION_MODES: dict[OperationMode, str] = {
+    OperationMode.GENERAL: "General mode",
```
The options in the state machine should be lowercase snake_case and preferably have a translation via a `translation_key`.

Example:

https://github.com/home-assistant/core/blob/32a62801390586178f7693b312c8f5bac3f86c74/homeassistant/components/plugwise/select.py#L51

https://github.com/home-assistant/core/blob/32a62801390586178f7693b312c8f5bac3f86c74/homeassistant/components/plugwise/strings.json#L44-L62

Note that changing the options to snake_case in the state machine is a breaking change.

mletenay replied:
Done, the states are snake_case and localized now.

mletenay committed:
Use proper and localized "operation mode" values

MartinHjelmare reviewed (approved):
Thanks!

Please add a list of how the state modes changed, ie old and new mode names, to the breaking change paragraph.

mletenay commented:
Changed select state values were added to breaking change paragraph.
Thanks for the review.

MartinHjelmare commented:
Thanks!
